src/dc_stock.rs

```diff
@@ -1,338 +1,112 @@
-use crate::constants::Event;
-use crate::context::Context;
-use crate::dc_contact::*;
-use crate::dc_tools::*;
-use crate::types::*;
-use crate::x::*;
-
-/* Return the string with the given ID by calling DC_EVENT_GET_STRING.
-The result must be free()'d! */
-pub unsafe fn dc_stock_str(context: &Context, id: libc::c_int) -> *mut libc::c_char {
-    return get_string(context, id, 0i32);
+/// Stock strings
+///
+/// These identify the string to return in [Context.stock_str].  The
+/// numbers must stay in sync with `deltachat.h` `DC_STR_*` constants.
+///
+/// See the `stock_*` methods on [Context] to use these.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, FromPrimitive, ToPrimitive)]
+#[repr(u32)]
+pub enum StringId {
+    NoMessages = 1,
+    SelfMsg = 2,
+    Draft = 3,
+    Member = 4,
+    Contact = 6,
+    VoiceMessage = 7,
+    DeadDrop = 8,
+    Image = 9,
+    Video = 10,
+    Audio = 11,
+    File = 12,
+    StatusLine = 13,
+    NewGroupDraft = 14,
+    MsgGrpName = 15,
+    MsgGrpImgChanged = 16,
+    MsgAddMember = 17,
+    MsgDelMember = 18,
+    MsgGroupLeft = 19,
+    Gif = 23,
+    EncryptedMsg = 24,
+    E2E_Available = 25,
+    Encr_Transp = 27,
+    Encr_None = 28,
+    CantDecrypt_Msg_Body = 29,
+    FingerPrints = 30,
+    ReadRcpt = 31,
+    ReadRcpt_MailBody = 32,
+    MsgGrpImgDeleted = 33,
+    E2E_Preferred = 34,
+    Contact_Verified = 35,
+    Contact_Not_Verified = 36,
+    Contact_Setup_Changed = 37,
+    ArchivedChats = 40,
+    StarredMsgs = 41,
+    AC_Setup_Msg_Subject = 42,
+    AC_Setup_Msg_Body = 43,
+    SelfTalk_SubTitle = 50,
+    Cannot_Login = 60,
+    Server_Response = 61,
+    MsgActionByUser = 62,
+    MsgActionByMe = 63,
+    MsgLocationEnabled = 64,
+    MsgLocationDisabled = 65,
+    Location = 66,
 }
 
-unsafe fn get_string(context: &Context, id: libc::c_int, qty: libc::c_int) -> *mut libc::c_char {
-    let mut ret: *mut libc::c_char;
-
-    ret =
-        context.call_cb(Event::GET_STRING, id as uintptr_t, qty as uintptr_t) as *mut libc::c_char;
-
-    if ret.is_null() {
-        ret = default_string(id)
-    }
-
-    ret
-}
-
-/* Add translated strings that are used by the messager backend.
-As the logging functions may use these strings, do not log any
-errors from here. */
-unsafe fn default_string(id: libc::c_int) -> *mut libc::c_char {
-    // TODO match on enum values /rtn
+/// Default untranslated strings for stock messages.
+///
+/// These could be used in logging calls, so no logging here.
+pub fn default_string(id: StringId) -> String {
```

Comment:
Two things:
- more rusty 

```rust
impl StringId {
    pub fn default_string(&self) -> String { ... }
}
```

- all these are static strings, so you can just change the signature to return `&'static str` avoiding the allocation of `String` in the method. The caller can still do that if they need to.